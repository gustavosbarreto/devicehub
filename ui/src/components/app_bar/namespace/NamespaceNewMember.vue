<template>
  <fragment>
    <v-btn
      class="ml-12"
      outlined
      data-test="dialog-btn"
      @click="dialog = !dialog"
    >
      Add Member
    </v-btn>

    <v-dialog
      v-model="dialog"
      max-width="450"
      @click:outside="close"
    >
      <v-card data-test="namespaceNewMember-dialog">
        <v-card-title class="headline grey lighten-2 text-center">
          Add member to Namespace
        </v-card-title>

        <ValidationObserver
          ref="obs"
          v-slot="{ passes }"
        >
          <v-card-text class="caption mb-0">
            <ValidationProvider
              v-slot="{ errors }"
              ref="providerUsername"
              vid="username"
              name="username"
              rules="required"
            >
              <v-text-field
                v-model="username"
                label="Username"
                :error-messages="errors"
                require
                data-test="username-text"
              />
            </ValidationProvider>
          </v-card-text>

          <v-card-actions>
            <v-spacer />
            <v-btn
              text
              data-test="close-btn"
              @click="close"
            >
              Close
            </v-btn>

            <v-btn
              color="primary"
              text
              data-test="add-btn"
              @click="passes(addMember)"
            >
              Add
            </v-btn>
          </v-card-actions>
        </ValidationObserver>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import {
  ValidationObserver,
  ValidationProvider,
} from 'vee-validate';

export default {
  name: 'NamespaceNewMemberComponent',

  components: {
    ValidationProvider,
    ValidationObserver,
  },

  props: {
    nsTenant: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      dialog: false,
      username: '',
    };
  },

  computed: {
    tenant() {
      return this.$props.nsTenant;
    },
  },

  methods: {
    async addMember() {
      try {
        await this.$store.dispatch('namespaces/addUser', {
          username: this.username,
          tenant_id: this.tenant,
        });

        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceNewMember);
        this.update();
      } catch (error) {
        if (error.response.status === 404) {
          this.$refs.obs.setErrors({
            username: 'The username doesn\'t exist.',
          });
        } else if (error.response.status === 409) {
          this.$refs.obs.setErrors({
            username: 'The username has already been added to namespace.',
          });
        } else {
          this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceNewMember);
        }
      }
    },

    update() {
      this.$emit('update');
      this.close();
    },

    close() {
      this.dialog = false;
      this.$refs.obs.reset();
      this.username = '';
    },
  },
};

</script>
