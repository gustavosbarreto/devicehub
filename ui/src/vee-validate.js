import {
  required,
  integer,
  email,
  confirmed,
} from 'vee-validate/dist/rules';
import { extend } from 'vee-validate';
import isValidHostname from 'is-valid-hostname';

import { parsePrivateKey, parseKey } from '@/sshpk';

extend('required', {
  ...required,
  message: 'This field is required',
});

extend('integer', {
  ...integer,
  message: 'This value must be a integer number',
});

extend('email', {
  ...email,
  message: 'This field must be a valid email',
});

extend('noDot', {
  validate: (value) => !/\./.test(value),
  message: 'The name must not contain dots',
});

extend('rfc1123', {
  validate: (value) => isValidHostname(value),
  message: 'You entered an invalid RFC1123 name',
});

extend('routeIdentifier', {
  validate: (value) => !/\/|@|&|:/.test(value),
  message: 'The name must not contain /, @, &, and :.',
});

extend('password', (value) => {
  if (value.length < 5 || value.length > 30) {
    return 'Your password should be 5-30 characters long';
  }
  return true;
});

extend('namespace', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your namespace should be 3-30 characters long';
  }
  return true;
});

extend('device', (value) => {
  if (value.length < 3 || value.length > 30) {
    return 'Your hostname should be 3-30 characters long';
  }
  return true;
});

extend('tag', (value) => {
  if (value.length < 3 || value.length > 255) {
    return 'Your tag should be 3-255 characters long';
  }
  return true;
});

extend('tagRequired', {
  ...required,
  message: 'You must choose at least one tag',
});

extend('tagsLength', {
  validate(value) {
    if (value.length > 3) {
      return 'The maximum capacity has reached';
    }
    return true;
  },
});

extend('comparePasswords', {
  validate(value, { currentPassword }) {
    if (value === currentPassword) {
      return false;
    }
    return true;
  },
  params: ['currentPassword'],
  message: 'The passwords are the same',
});

extend('confirmed', {
  ...confirmed,
  message: 'The passwords do not match',
});

extend('parseKey', {
  validate(value, args) {
    try {
      if (args.typeKey === 'private') {
        parsePrivateKey(value);
      } else {
        parseKey(value);
      }

      return true;
    } catch (err) {
      return false;
    }
  },
  params: ['typeKey'],
  message: 'Not valid key',
});
