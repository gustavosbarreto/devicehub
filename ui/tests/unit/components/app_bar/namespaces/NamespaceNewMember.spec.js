import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import NamespaceNewMember from '@/components/app_bar/namespace/NamespaceNewMember';
import { ValidationProvider, ValidationObserver } from 'vee-validate';
import flushPromises from 'flush-promises';
import Vuetify from 'vuetify';
import '@/vee-validate';

describe('NamespaceNewMember', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  document.body.setAttribute('data-app', true);
  localVue.use(Vuex);
  localVue.component('ValidationProvider', ValidationProvider);
  localVue.component('ValidationObserver', ValidationObserver);

  let wrapper;

  const tenant = 'xxxxx';

  const store = new Vuex.Store({
    namespaced: true,
    actions: {
      'namespaces/adduser': () => {
      },
      'snackbar/showSnackbarSuccessAction': () => {
      },
      'snackbar/showSnackbarErrorAction': () => {
      },
    },
  });

  ///////
  // In this case, the rendering of the button is checked.
  ///////

  describe('Button', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceNewMember, {
        store,
        localVue,
        stubs: ['fragment'],
        propsData: { nsTenant: tenant },
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receive data in props', () => {
      expect(wrapper.vm.nsTenant).toEqual(tenant);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(false);
      expect(wrapper.vm.username).toEqual('');
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.tenant).toEqual(tenant);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="add-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="namespaceNewMember-dialog"]').exists()).toEqual(false);
    });
  });

  ///////
  // In this case, the focus of the test is dialog rendering.
  ///////

  describe('Button', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceNewMember, {
        store,
        localVue,
        stubs: ['fragment'],
        propsData: { nsTenant: tenant },
        vuetify,
      });

      wrapper.setData({ dialog: true });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receive data in props', () => {
      expect(wrapper.vm.nsTenant).toEqual(tenant);
    });
    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(true);
      expect(wrapper.vm.username).toEqual('');
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.tenant).toEqual(tenant);
    });

    //////
    // HTML validation
    //////

    it('Show validation messages', async () => {
      wrapper.setData({ namespaceName: '' });
      await flushPromises();

      const validator = wrapper.vm.$refs.providerUsername;

      await validator.validate();
      expect(validator.errors[0]).toBe('This field is required');
    });
    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="add-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="namespaceNewMember-dialog"]').exists()).toEqual(true);
    });
  });
});
