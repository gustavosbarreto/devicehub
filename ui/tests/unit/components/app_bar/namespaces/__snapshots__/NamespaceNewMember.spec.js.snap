// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceNewMember Button Renders the component 1`] = `
<fragment-stub><button type="button" class="ml-12 v-btn v-btn--outlined theme--light v-size--default" data-test="dialog-btn"><span class="v-btn__content">
    Add Member
  </span></button>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceNewMember Dialog Renders the component 1`] = `
<fragment-stub><button type="button" class="ml-12 v-btn v-btn--outlined theme--light v-size--default" data-test="dialog-btn"><span class="v-btn__content">
    Add Member
  </span></button>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;
